[CUBRIDMAN-333] Add descriptions for redesigned connection layer and elastic worker pool. (#734)
CUBRID/cubrid-manual: en/admin/config.rst

      cubrid-dev1
      CUB2.dev

.. _max_clients:

**max_clients**

    **max_clients** is a parameter to configure the maximum number of clients (usually broker application processes (CAS)) which allow concurrent connections to the database server. The **max_clients** parameter refers to the number of concurrent transactions per database server process. The default value is **100**.

>     To guarantee performance while increasing the number of concurrent users in CUBRID environment, you need to make the appropriate value of the **max_clients** (**cubrid.conf**) parameter and the :ref:`MAX_NUM_APPL_SERVER <max-num-appl-server>` (**cubrid_broker.conf**) parameter. That is, you are required to configure the number of concurrent connections allowed by databases with the **max_clients** parameter. You should also configure the number of concurrent connections allowed by brokers with the **MAX_NUM_APPL_SERVER** parameter.

Reviewer: ```suggestion
    To guarantee performance while increasing the number of concurrent users in the CUBRID environment, you need to set appropriate values for the **max_clients** (**cubrid.conf**) parameter and the :ref:`MAX_NUM_APPL_SERVER <max-num-appl-server>` (**cubrid_broker.conf**) parameter. That is, you are required to configure the number of concurrent connections allowed by the database with the **max_clients** parameter. You should also configure the number of concurrent connections allowed by brokers with the **MAX_NUM_APPL_SERVER** parameter.
```

>     For example, in the **cubrid_broker.conf** file, two node of a broker where the **MAX_NUM_APPL_SERVER** value of [%query_editor] is 50 and the **MAX_NUM_APPL_SERVER** value of [%BROKER1] is 50 is trying to connect one database server, the concurrent connections (**max_clients** value) allowed by the database server can be configured as follows:

Reviewer: ```suggestion
    For example, in the **cubrid_broker.conf** file, if two broker nodes, [%query_editor] and [%BROKER1], each have a **MAX_NUM_APPL_SERVER** value of 50 and connect to one database server, the concurrent connections (the **max_clients** value) allowed by the database server can be configured as follows:
```

>     *   (the maximum number of 50 by each node of a broker) * (two node of a broker) + (10 spare for database server connections of internal CUBRID process such as database server connection of CSQL Interpreter or HA log replication process) = 110

Reviewer: ```suggestion
    *   (maximum of 50 for each broker node) * (two broker nodes) + (10 spare database server connections for internal CUBRID processes, such as the CSQL Interpreter or HA log replication) = 110
```

>     Especially, in HA environment, the value must be greater than the sum specified in **MAX_NUM_APPL_SERVER** of every broker node which connects to the same database.

Reviewer: ```suggestion
    In particular, in HA environments, the value must be greater than the sum specified in **MAX_NUM_APPL_SERVER** of every broker node that connects to the same database.
```

>     Note that the memory usage is affected by the value specified in **max_clients**. That is, if the number of value is high, the memory usage will increase regardless of whether or not the clients actually access the database.

Reviewer: ```suggestion
    Note that memory usage is affected by the value specified in **max_clients**. That is, if the value is high, memory usage will increase regardless of whether clients actually access the database.
```

    .. note::
        

>         In Linux system, max_clients parameter is related to "ulimit -n" command, which specifies the maximum number of file descriptors which a process can use. File descriptor includes not only a file, but also a network socket. Therefore, the number of "ulimit -n" should be greater than the number of max_clients.

Reviewer: ```suggestion
        In Linux systems, the **max_clients** parameter is related to the "ulimit -n" command, which specifies the maximum number of file descriptors that a process can use. File descriptors include not only files, but also network sockets. Therefore, the value of "ulimit -n" should be greater than the value of **max_clients**.
```

**auto_scaling_window_size**

    **auto_scaling_window_size** specifies the search range used by the server when finding an appropriate number of **connection workers**. If **auto_scaling_window_size** is too large, performance may decrease. If it is too small, proper adjustment may be difficult.

.. _budget_per_connection:

**recv_budget_per_connection**

    **recv_budget_per_connection** specifies the number of bytes one connection can receive at once. If **recv_budget_per_connection** is too small, receive processing may be interrupted too frequently and become slower. If it is too large, receives of other connections may be delayed.

**send_budget_per_connection**

    **send_budget_per_connection** specifies the number of bytes one connection can send at once. If **send_budget_per_connection** is too small, send processing may be interrupted too frequently and become slower. If it is too large, sends of other connections may be delayed.

.. _tcp_keepalive:

**tcp_keepalive**
  

>     **tcp_keepalive** is a parameter which specifies if you apply SO_KEEPALIVE option to TCP network protocol or not. The default is **yes**. If this value is **no**, DB server-side connection can be disconnected when transaction logs are not copied for a long time in the firewall environment between master and slave.

Reviewer: ```suggestion
    **tcp_keepalive** is a parameter that specifies whether to apply the SO_KEEPALIVE option to the TCP network protocol. The default is **yes**. If this value is **no**, the DB server-side connection can be disconnected when transaction logs are not copied for a long time in a firewall environment between master and slave.
```

+---------------------------------------+--------+-------------------+---------+-----------------+
| thread_worker_timeout_seconds         | int    | 300               | -1      | 3600            |
+---------------------------------------+--------+-------------------+---------+-----------------+
| loaddb_worker_count                   | bool   | 8                 | 2       | 64              |
+---------------------------------------+--------+-------------------+---------+-----------------+

**task_group**

    **task_group** adjusts the number of **task worker** groups. The default value is set to the number of system cores.

**task_worker**

    **task_worker** specifies the number of **workers** used for transaction jobs. The default value is the same as **max_clients**. **task_worker** has a significant impact on performance. If set too high, performance may degrade due to frequent entry into critical sections and higher resource usage. If set too low, resources may not be fully utilized, or processing may stall when deadlocks occur.

    The optimal value of **task_worker** depends on the environment, so tune it carefully.

**thread_worker_pooling**

>     If **thread_worker_pooling** parameter is true, all threads used for client requests execution are pooled on server boot.

Reviewer: ```suggestion
    If the **thread_worker_pooling** parameter is true, all threads used for client request execution are pooled when the server boots.
```
